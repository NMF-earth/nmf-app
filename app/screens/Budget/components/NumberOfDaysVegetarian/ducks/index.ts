import selectors from "./NumberOfDaysVegetarian.selectors";

export { selectors };
