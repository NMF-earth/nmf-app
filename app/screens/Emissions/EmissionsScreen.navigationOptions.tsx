import React from "react";

import { Text } from "../../components";
import { t } from "../../utils";
import { Layout, ComponentsStyle } from "../../style";

const navigationOptions = () => ({
  headerStyle: {
    ...ComponentsStyle.header,
  },
  headerBackTitle: null,
  headerTitle: () => (
    <Text.H1 style={Layout.androidNavTitle}>
      {t("EMISSIONS_SCREEN_TITLE")}
    </Text.H1>
  ),
});

export default navigationOptions;
