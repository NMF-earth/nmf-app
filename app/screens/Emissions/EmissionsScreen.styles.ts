import { StyleSheet } from "react-native";

import { PADDING_HORIZONTAL } from "../../constants/Layout";
import { Layout } from "../../style";

const button = {
  flex: 1,
  marginBottom: 20,
};
export default StyleSheet.create({
  container: {
    ...Layout.containerNoPadding,
  },
  buttonView: {
    flexDirection: "row",
    position: "absolute",
    bottom: 0,
    right: 0,
    left: 0,
  },
  buttonLeft: {
    ...button,
    marginLeft: PADDING_HORIZONTAL,
    marginRight: PADDING_HORIZONTAL / 2,
  },
  buttonRight: {
    ...button,
    marginRight: PADDING_HORIZONTAL,
    marginLeft: PADDING_HORIZONTAL / 2,
  },
  textNoEmission: {
    paddingVertical: 20,
  },
  separator: {
    height: 100,
  },
});
