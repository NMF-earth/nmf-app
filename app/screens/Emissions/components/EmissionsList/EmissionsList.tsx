import React from "react";
import { SectionList, View } from "react-native";
import { useNavigation } from "@react-navigation/native";

import { navigate } from "navigation";

import EmissionsListItem from "../EmissionsListItem";
import EmissionsListItemProps from "../EmissionsListItem/EmissionsListItemProps";
import SectionHeader from "../SectionHeader";
import styles from "./EmissionsList.styles";

const EmissionsList = ({ emissions, monthlyCarbonBudget }) => {
  const navigation = useNavigation();
  const navigator = navigate(navigation);

  const renderListFooter = () => <View style={styles.separator} />;
  const renderSectionHeader = (date, co2value) => (
    <SectionHeader
      date={date}
      co2value={co2value}
      monthlyCarbonBudget={monthlyCarbonBudget}
    />
  );

  return (
    <SectionList<EmissionsListItemProps>
      sections={emissions}
      stickySectionHeadersEnabled
      ListFooterComponent={renderListFooter}
      renderSectionHeader={({ section: { date, co2value } }) =>
        renderSectionHeader(date, co2value)
      }
      keyExtractor={({ id }) => id}
      renderItem={({
        item: { id, isMitigated, name, title, co2value, iconName },
      }) => (
        <EmissionsListItem
          id={id}
          isMitigated={isMitigated}
          name={name}
          onPress={() => navigator.openEmissionItem({ id })}
          title={title}
          co2value={co2value}
          iconName={iconName}
        />
      )}
    />
  );
};

export default EmissionsList;
