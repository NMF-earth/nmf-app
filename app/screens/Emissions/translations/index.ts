import en from "./en";
import fr from "./fr";
import de from "./de";
import sv from "./sv";

interface TranslationKeys {
  EMISSIONS_SCREEN_TAB_NAME: string;
  EMISSIONS_SCREEN_TITLE: string;
  EMISSIONS_SCREEN_TO_OFFSET: string;
  EMISSIONS_SCREEN_MITIGATED: string;
  EMISSIONS_SCREEN_MITIGATE: string;
  EMISSIONS_SCREEN_ADD: string;
  EMISSIONS_SCREEN_NO_EMISSION_MITIGATED: string;
}

export { en, fr, de, sv, TranslationKeys };
