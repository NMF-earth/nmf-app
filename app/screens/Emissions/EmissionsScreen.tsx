import React from "react";
import { FlatList, SafeAreaView, View } from "react-native";
import { useSelector } from "react-redux";
import { t } from "../../utils";
import styles from "./EmissionsScreen.styles";
import {
  TabbedView,
  Text,
  EmissionListItem,
  EmissionListItemProps
} from "../../components";
import { AddEmissionAndMitigateButtons } from "./components";
import { selectors } from "./ducks";
import { navigate } from "../../navigation";

const EmissionsScreen = props => {
  const navigator = navigate(props.navigation);
  const emissionsToMitigate = useSelector(selectors.getEmissionsToMitigate);
  const emissionsMitigated = useSelector(selectors.getEmissionsMitigated);

  return (
    <React.Fragment>
      <SafeAreaView style={styles.container}>
        <TabbedView
          items={[
            {
              title: t("EMISSIONS_SCREEN_TO_OFFSET"),
              component: (
                <FlatList<EmissionListItemProps>
                  ListFooterComponent={<View style={styles.separator} />}
                  style={styles.listContainer}
                  data={emissionsToMitigate}
                  keyExtractor={item => item.id}
                  renderItem={({ item }: { item: EmissionListItemProps }) => (
                    <EmissionListItem
                      id={item.id}
                      name={item.name}
                      onPress={() =>
                        navigator.openEmissionItem({ id: item.id })
                      }
                      title={item.title}
                      co2value={item.co2value}
                      food={item.food}
                      transport={item.transport}
                      custom={item.custom}
                    />
                  )}
                />
              )
            },
            {
              title: t("EMISSIONS_SCREEN_MITIGATED"),
              component: emissionsMitigated.length ? (
                <FlatList<EmissionListItemProps>
                  ListFooterComponent={<View style={styles.separator} />}
                  style={styles.listContainer}
                  data={emissionsMitigated}
                  keyExtractor={item => item.id}
                  renderItem={({ item }: { item: EmissionListItemProps }) => (
                    <EmissionListItem
                      id={item.id}
                      name={item.name}
                      onPress={item.onPress}
                      title={item.title}
                      co2value={item.co2value}
                      food={item.food}
                      transport={item.transport}
                      custom={item.custom}
                    />
                  )}
                />
              ) : (
                <Text.Primary style={styles.textNoEmission} center lightGray>
                  {t("EMISSIONS_SCREEN_NO_EMISSION_MITIGATED")}
                </Text.Primary>
              )
            }
          ]}
        />
      </SafeAreaView>
      <AddEmissionAndMitigateButtons navigation={props.navigation} />
    </React.Fragment>
  );
};

export default EmissionsScreen;
