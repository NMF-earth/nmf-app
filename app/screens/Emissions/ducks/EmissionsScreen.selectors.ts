import { map, pipe, groupBy, toPairs, sum } from "ramda";
import moment from "moment";

import { emissions } from "ducks";

import { Emission } from "../../../interfaces";
import { calculation, ui } from "../../../utils";

interface EmissionListItem extends Emission {
  title: string;
  co2value: number;
  iconName: string;
  onPress: () => void;
}

const getEmissionListItem = (item: Emission) => {
  const emissionItem: EmissionListItem = {
    ...item,
    title: ui.getTranslationModelType(item.emissionModelType),
    co2value: calculation.getC02ValueFromEmission(item),
    iconName: ui.getIconFromModelType(item.emissionModelType),
    onPress: () => {
      // do nothing.
    },
  };

  return emissionItem;
};

const getStartOfMonth = (time) => moment(time).startOf("month").format();

const groupByMonth = groupBy((item: EmissionListItem) =>
  getStartOfMonth(item.creationDate)
);

const dateObjMap = map(([date, data, co2value]) => ({
  date: date,
  data: data,
  co2value: co2value,
}));

/* moment().utc().toISOString() gives "YYYY-MM-DDTHH:mm:ss.sssZ" */
const filterByMostRecent = (array: [EmissionListItem]) =>
  array.sort((a, b) => +new Date(b.creationDate) - +new Date(a.creationDate));

const getMonthlyPourcentage = (items) =>
  map(
    (item) => [...item, sum(map((emission) => emission.co2value, item[1]))],
    items
  );

const getEmissions = (state) =>
  pipe(
    emissions.selectors.getAllEmissions,
    map(getEmissionListItem),
    filterByMostRecent,
    groupByMonth,
    toPairs,
    getMonthlyPourcentage,
    dateObjMap
  )(state);

export default {
  getEmissions,
  getEmissionListItem,
};
