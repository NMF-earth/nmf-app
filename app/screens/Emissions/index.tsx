import React from "react";
import { useSelector } from "react-redux";
import NoEmission from "../../components/NoEmission";
import EmissionsScreen from "./EmissionsScreen";
import navigationOptions from "./EmissionsScreen.navigationOptions";
import { EmissionListItemProps } from "../../components";
import { emissions } from "../../ducks";

interface Props {
  emissionsToMitigate: Array<EmissionListItemProps>;
  emissionsMitigated: Array<EmissionListItemProps>;
  navigation: {
    push: (screen: string) => void;
  };
}

const Emissions = (props: Props) => {
  const emissionsToMitigate = useSelector(
    emissions.selectors.getEmissionsToMitigate
  );
  const emissionsMitigated = useSelector(
    emissions.selectors.getEmissionsMitigated
  );
  const { navigation } = props;

  if (emissionsToMitigate.length || emissionsMitigated.length) {
    return <EmissionsScreen navigation={navigation} />;
  }

  return <NoEmission addEmission={() => navigation.push("AddEmission")} />;
};

Emissions.navigationOptions = navigationOptions;

export default Emissions;
