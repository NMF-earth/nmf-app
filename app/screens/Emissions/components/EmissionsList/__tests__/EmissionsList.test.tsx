import React from "react";
import { create } from "react-test-renderer";

import { FoodEnum, TransportEnum } from "carbon-footprint";

import emissions from "ducks/emissions";

import { selectors } from "../../../../../screens/Emissions/ducks";

import EmissionsList from "../EmissionsList";
import {
  Emission as EmissionType,
  EmissionEnum,
} from "../../../../../interfaces";

jest.unmock("../EmissionsList");

const emissionNotMitigatedOld: EmissionType = {
  id: "3",
  creationDate: new Date("2020-01-01T12:08:16.623Z").toISOString(),
  emissionModelType: FoodEnum.beans,
  emissionType: EmissionEnum.food,
  isMitigated: false,
  value: 30,
};

const emissionNotMitigated: EmissionType = {
  id: "1",
  creationDate: new Date("2020-03-01T12:08:16.623Z").toISOString(),
  emissionModelType: FoodEnum.beans,
  emissionType: EmissionEnum.food,
  isMitigated: false,
  value: 10,
};

const emissionMitigated: EmissionType = {
  id: "12",
  creationDate: new Date("2020-03-01T12:08:16.623Z").toISOString(),
  emissionModelType: TransportEnum.boat,
  emissionType: EmissionEnum.transport,
  isMitigated: true,
  value: 100,
};

const state = {
  [emissions.namespace]: [
    emissionNotMitigated,
    emissionMitigated,
    emissionNotMitigatedOld,
  ],
};

const props = {
  emissions: selectors.getEmissions(state),
  monthlyCarbonBudget: 24,
  navigator: {
    push: () => {
      // do nothing.
    },
    navigate: () => {
      // do nothing.
    },
  },
};

it("EmissionsList renders correctly", () => {
  const tree = create(<EmissionsList {...props} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("EmissionsList renders correctly if no emissions", () => {
  const tree = create(<EmissionsList {...props} emissions={[]} />).toJSON();
  expect(tree).toMatchSnapshot();
});
