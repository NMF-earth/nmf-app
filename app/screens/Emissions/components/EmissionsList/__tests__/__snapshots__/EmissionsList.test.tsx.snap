// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmissionsList renders correctly 1`] = `
<RCTScrollView
  ListFooterComponent={[Function]}
  data={
    Array [
      Object {
        "co2value": 20.0803,
        "data": Array [
          Object {
            "co2value": 20,
            "creationDate": "2020-03-01T12:08:16.623Z",
            "emissionModelType": "beans",
            "emissionType": "food",
            "iconName": "md-build",
            "id": "1",
            "isMitigated": false,
            "onPress": [Function],
            "title": "Custom",
            "value": 10,
          },
          Object {
            "co2value": 0.0803,
            "creationDate": "2020-03-01T12:08:16.623Z",
            "emissionModelType": "boat",
            "emissionType": "transport",
            "iconName": "md-boat",
            "id": "12",
            "isMitigated": true,
            "onPress": [Function],
            "title": "Boat",
            "value": 100,
          },
        ],
        "date": "2020-03-01T00:00:00Z",
      },
      Object {
        "co2value": 60,
        "data": Array [
          Object {
            "co2value": 60,
            "creationDate": "2020-01-01T12:08:16.623Z",
            "emissionModelType": "beans",
            "emissionType": "food",
            "iconName": "md-build",
            "id": "3",
            "isMitigated": false,
            "onPress": [Function],
            "title": "Custom",
            "value": 30,
          },
        ],
        "date": "2020-01-01T00:00:00Z",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={
    Array [
      0,
      4,
    ]
  }
  updateCellsBatchingPeriod={50}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <SectionHeader
        co2value={20.0803}
        date="2020-03-01T00:00:00Z"
        monthlyCarbonBudget={24}
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <EmissionsListItem
        co2value={20}
        iconName="md-build"
        id="1"
        isMitigated={false}
        onPress={[Function]}
        title="Custom"
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <EmissionsListItem
        co2value={0.0803}
        iconName="md-boat"
        id="12"
        isMitigated={true}
        onPress={[Function]}
        title="Boat"
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    />
    <View
      onLayout={[Function]}
      style={null}
    >
      <SectionHeader
        co2value={60}
        date="2020-01-01T00:00:00Z"
        monthlyCarbonBudget={24}
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <EmissionsListItem
        co2value={60}
        iconName="md-build"
        id="3"
        isMitigated={false}
        onPress={[Function]}
        title="Custom"
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    />
    <View
      onLayout={[Function]}
    >
      <View
        style={
          Object {
            "height": 100,
          }
        }
      />
    </View>
  </View>
</RCTScrollView>
`;

exports[`EmissionsList renders correctly if no emissions 1`] = `
<RCTScrollView
  ListFooterComponent={[Function]}
  data={Array []}
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
    >
      <View
        style={
          Object {
            "height": 100,
          }
        }
      />
    </View>
  </View>
</RCTScrollView>
`;
